pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/QueryRunner.java

```diff
@@ -139,23 +143,26 @@ public void processQuery(DistributedStagePlan distributedStagePlan, ExecutorServ
     }
   }
 
-  private BaseDataBlock processServerQuery(ServerQueryRequest serverQueryRequest,
+  private TransferableBlock processServerQuery(ServerQueryRequest serverQueryRequest,
       ExecutorService executorService) {
-    BaseDataBlock dataBlock;
-    try {
-      DataTable dataTable = _serverExecutor.processQuery(serverQueryRequest, executorService, null);
-      if (!dataTable.getExceptions().isEmpty()) {
-        // if contains exception, directly return a metadata block with the exceptions.
-        dataBlock = DataBlockUtils.getErrorDataBlock(dataTable.getExceptions());
-      } else {
-        // this works because default DataTableImplV3 will have a version number at beginning:
-        // the new DataBlock encodes lower 16 bits as version and upper 16 bits as type (ROW, COLUMNAR, METADATA)
-        dataBlock = DataBlockUtils.getDataBlock(ByteBuffer.wrap(dataTable.toBytes()));
+    BaseResultsBlock resultsBlock = _serverExecutor.processQueryV2(serverQueryRequest, executorService, null);
+    if (CollectionUtils.isNotEmpty(resultsBlock.getProcessingExceptions())) {
+      final Map<Integer, String> exceptions = new HashMap<>();
+      resultsBlock.getProcessingExceptions().forEach(x -> exceptions.put(x.getErrorCode(), x.getMessage()));
+      return new TransferableBlock(DataBlockUtils.getErrorDataBlock(exceptions));
+    }
+    if (resultsBlock instanceof SelectionResultsBlock) {
+      SelectionResultsBlock selectionResultsBlock = (SelectionResultsBlock) resultsBlock;
+      return new TransferableBlock((List<Object[]>) selectionResultsBlock.getRows(),
+          selectionResultsBlock.getDataSchema(), BaseDataBlock.Type.ROW);
+    } else {
+      try {
```

Comment:
The conversions below are similar to what was happening before. At present I was only able to optimize `SelectionResultsBlock` conversion.